modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/BinaryTupleBuilder.java

```diff
@@ -315,8 +315,19 @@ public BinaryTupleBuilder appendNumber(BigInteger value) {
      * @return {@code this} for chaining.
      */
     public BinaryTupleBuilder appendDecimalNotNull(BigDecimal value, int scale) {
-        // TODO IGNITE-21745: Inefficient serialization, many bytes wasted to store small values when scale is big.
-        putBytes(value.setScale(scale, RoundingMode.HALF_UP).unscaledValue().toByteArray());
+        if (value.scale() > scale) {
+            value = value.setScale(scale, RoundingMode.HALF_UP);
+        }
+
+        value = value.stripTrailingZeros();
+
+        // See CatalogUtils.MAX_DECIMAL_SCALE = Short.MAX_VALUE
+        assert value.scale() <= Short.MAX_VALUE : "Decimal scale is too large: " + value.scale() + " > " + Short.MAX_VALUE;
+        assert value.scale() >= Short.MIN_VALUE : "Decimal scale is too small: " + value.scale() + " < " + Short.MIN_VALUE;
```

Comment:
Can this actually happen if user uses such values? Maybe, we should use exceptions here instead? Let's maybe add tests with such a values.

Reply:
Agree. Replaced with exceptions, added test.